### ISSUE TITLE
This discussion starts from this PR: https://github.com/OpenRLHF/OpenRLHF/pull/477

### ISSUE THREAD

**Issue by ZetangForward** [OpenRLHF/OpenRLHF#489]
This discussion starts from this PR: https://github.com/OpenRLHF/OpenRLHF/pull/477

## Discussion 1: loss accumulation

I noticed a very important thing: **the loss accumulation logic seems strange in the raw code**:
https://github.com/OpenRLHF/OpenRLHF/blame/5f52e354829e402561313a8c56e0a5c071a3ae65/openrlhf/trainer/sft_trainer.py#L164

In fact, the normal logic for gradient accumulation should 
1. involve the loss being continuously accumulated through `self.strategy.backward`, and then updating the optimizer when `step % self.strategy.accumulated_gradients` is reached; 
2. At the same time, the gradient (loss) for each step should be divided by `accumulated_gradients` rather than directly propagating backward as in the original code

Here's a code snippet that illustrates the correct way to handle gradient accumulation:
```python
# Assuming self.strategy.accumulated_gradients is the number of steps to accumulate gradients
# and self.optimizer is the optimizer instance.

for step, batch in enumerate(train_loader):
    optimizer.zero_grad()
    outputs = model(batch)
    loss = criterion(outputs, targets)
    loss = loss / self.strategy.accumulated_gradients  # Normalize the loss
    self.strategy.backward(loss)  # Accumulate gradients
    if (step + 1) % self.strategy.accumulated_gradients == 0:
        optimizer.step()  # Update the optimizer
        optimizer.zero_grad()  # Reset gradients for the next accumulation cycle
```
This code snippet ensures that the gradients are accumulated over multiple steps and then normalized before updating the optimizer. This approach should help in reducing the discrepancy you observed when using ring attention.

Here is the original code for comparison:
```python
...

loss = ... # CE loss or other loss
self.strategy.backward(loss, self.model, self.optimizer)  # backward the current loss
self.strategy.optimizer_step(self.optimizer, self.model, self.scheduler)  # update the optimizer (no accumulation)

step_bar.update()

loss_mean = loss_mean * 0.9 + 0.1 * gpt_loss.item()
logs_dict = {
    "gpt_loss": gpt_loss.item(),  # here is the mini batch loss not the accumulated loss
    "loss_mean": loss_mean,
    "lr": self.scheduler.get_last_lr()[0],
}

...

# step bar
logs_dict = self.strategy.all_reduce(logs_dict)  # collect all the mini batch loss
step_bar.set_postfix(logs_dict)
step_bar.update()

# logs/checkpoints/evaluation
if step % self.strategy.accumulated_gradient == 0:  # just update the log not the loss
    global_step = step // self.strategy.accumulated_gradient
    client_states = {"consumed_samples": global_step * args.train_batch_size}
    self.save_logs_and_checkpoints(args, global_step, step_bar, logs_dict, client_states)
```

When I tested, I found that the original code would cause a significant deviation when ring attention is turned on or off. 
The reason is that _when ring attention is turned on, there will inevitably be gradient accumulation (because a batch of data is not as much as vanilla attention), but using the original code would result in "fake gradient accumulation."

## Discussion 2: Drop_last of the Dataloader

The data loader currently drops the last samples based on the micro-batch size, not the training batch size. 

https://github.com/OpenRLHF/OpenRLHF/blob/5f52e354829e402561313a8c56e0a5c071a3ae65/openrlhf/cli/train_sft.py#L78

This means that if the total number of samples in an epoch is not perfectly divisible by the micro-batch size, **the remaining samples aren't dropped.  Consequently, these leftover samples might be merged with the new batch at the start of the next epoch.** 

If gradient accumulation is not properly reset at the beginning of each epoch, this could lead to "carryover" effects where the gradients from the previous epoch influence the calculations in the new epoch.

In my modified code, the result is the accumulation of gradients, which is why the vanilla loss and ring attention loss curves in wandb basically overlap (see https://wandb.ai/zecheng-tang/debug_openrlhf_train_sft/workspace?nw=nwuserzechengtang in PR: https://github.com/OpenRLHF/OpenRLHF/pull/477).

I'm not sure if my understanding is correct or if the DeepSpeed within the OpenRLHF library automatically handles the above operations (but upon reviewing the code, it seems like the logic mentioned above is not present). If there are any mistakes, please correct me.

_Originally posted by @ZetangForward in https://github.com/OpenRLHF/OpenRLHF/discussions/481_
